tests/TESTING.md

```diff
@@ -0,0 +1,105 @@
+See the [etcd unit test](https://github.com/k3s-io/k3s/blob/master/pkg/etcd/etcd_test.go) as an example.
+
+### Running
+
+```bash
+go test ./pkg/... -run Unit
+```
+
+Note that as unit tests call functions directly, they are the primary drivers of K3s's code coverage
+metric.
+
+___
+
+## Integration Tests
+
+Integration tests should be used to test a specific functionality of k3s that exists across multiple go packages, either via exported function calls, or more often, CLI comands.
+Integration tests should be used for "black box" testing. 
+
+### Framework
+
+All integration tests in K3s follow a [Behavior Diven Development (BDD)](https://en.wikipedia.org/wiki/Behavior-driven_development) style. Specifically, K3s uses [Ginkgo](https://onsi.github.io/ginkgo/) and [Gomega](https://github.com/onsi/gomega) to drive the tests.  
+To generate an initial test, the command `ginkgo bootstrap` can be used.
+
+To facilitate K3s CLI testing, see `tests/testutil/cmd.go` helper functions.
+
+### Format
+
+Integration tests can be placed in two areas:  
+
+1. Next to the go package they intend to test.
+2. In `tests/integration/` for package agnostic testing.  
+
+Package specific integration tests should use the `<PACKAGE_UNDER_TEST>_test` package.  
+Package agnostic integration tests should use the `integration` package.  
+All integration test files should be named: `<TEST_NAME>_int_test.go`  
+All integration test functions should be named: `Test_Integration<Test_Name>`.  
+See the [etcd snapshot test](https://github.com/k3s-io/k3s/blob/master/pkg/etcd/etcd_int_test.go) as a package specific example.  
+See the [local storage test](https://github.com/k3s-io/k3s/blob/master/tests/integration/localstorage_int_test.go) as a package agnostic example.
+
+### Running
+
+```bash
+go test ./pkg/... ./tests/... -run Integration
+```
+
+___
+
+## End-to-End (E2E) Tests
+
+End-to-end tests utilize Ginkgo and Gomega like the integration tests, but rely on separate testing utilities and are self-contained within the `test/e2e` directory. E2E tests cover complete K3s single and multi-cluster configuration and administration: bringup, update, teardown etc.  
+E2E tests are run nightly as part of K3s quality assurance (QA).
+
+### Format
+
+All E2E tests should be placed under the `e2e` package.  
+All E2E test functions should be named: `Test_E2E<TEST_NAME>`.  
+See the [upgrade cluster test](https://github.com/k3s-io/k3s/blob/master/tests/e2e/upgradecluster_test.go) as an example.
+
+### Running
+
+Generally, E2E tests are run as a nightly Jenkins job for QA. They can still be run locally but additional setup may be required.
+
+```bash
+go test ./tests/... -run E2E
+```
+
+## Contributing New Tests
+
+___
+We gladly accept new tests of all types. If you wish to create
+a new test, please submit a PR with a title that includes the words `<NAME_OF_TEST> Created`. Additionally, the final merge commit into K3s should also include those keywords. This
```

Comment:
We may want to consider creating a new label for these issues/PR's to be easier to identify or report on. /cc @cwayne18

Reply:
That would be a good idea. These rules for title and commits come from Shylaja on how QA wants to process and track testing.

Reply:
We may also then want to include a link to documentation that outlines those guidelines.

Reply:
Holding off on submitting PR until next design call on this issue.